### DISCUSSION TITLE
WinKeyer Serial Protocol

### DISCUSSION THREAD

**Discussion by jimahlstrom** [softerhardware/CWKeyer#41] [General]
The WinKeyer has a serial port interface that is used to control the keyer and to connect to loggers and contest programs. Many other keyers such as K3NG support this protocol, and there is a whole ecosystem of connected programs that use it. I think it is important that our CWKeyer support this serial protocol if we expect it to be used by serious CW operators.

The protocol has a couple dozen commands. Almost all are sent from the PC to the keyer. There are only four from the keyer to the PC, such as changes to the speed pot.

The WinKeyer can run in either host mode or stand-alone mode. In host mode the PC controls the keyer over the serial port. But since serial ports are exclusive, AFAIK only one program can connect. This could be a keyer control program or a logger or contest program. In stand-alone mode the keyer runs on stored settings plus its speed pot and buttons and no PC is required for operation. There could still be a logger or contest program attached. The settings for stand-alone mode are separate from the settings for host mode and must be programmed in advance. WinKeyer supplies a Windows program for this. There must be a program for Linux and Mac but I have not been able to find it. I am willing to write one.

Our CWKeyer already has a serial port, so all we need is the protocol at [https://hamcrafters2.com/files/WK2_Datasheet_v23.pdf](https://hamcrafters2.com/files/WK2_Datasheet_v23.pdf) and the ability to write stand-alone settings to flash memory.

I have been thinking about how to support the CWKeyer from Quisk. Quisk already has a Midi port to control its GUI. But Midi ports are exclusive, so it will need a second Midi port for the CWKeyer. I never thought that writing a whole page of keyer setting into Quisk was a good idea. This is a separate application that should run in its own window. But even that is unnecessary. There are already programs to control the keyer over the serial port. We can just use those. All Quisk needs is a few Midi messages for CW, PTT and speed change.

**Comment by dl1ycf** [comment]
Winkey control programs are available. For the Mac, there is "MacWinkeyeApp", https://github.com/psichel/MacWinKeyer,
and for Linux (but it works also for MacOS) there is flwkey, https://sourceforge.net/p/fldigi/flwkey/ci/master/tree/.

I worked a little bit on flwkey, some error corrections and the possiblity to read/write EEPROM settings, you find this on
my (dl1ycf) github page.

I am not 100% sure, but I think the EEPROM layout of the K3NG keyer is not compatible (I have to try this again) so the settings cannot be set the usual way, but I must check. If it turns out this way, it is a serious drawback.
EDITED LATER: but I think the k3ng writes the current settings into EEPROM regularly, so you only have to change the settings in host mode and they then should apply in standalone mode. I have not run k3ng for some time know, all these statements need verification.

**Reply by jimahlstrom** [reply]
I tried your version of flwkey on Ubuntu. There is no "configure" file, and creating one with autoconf gives errors: 
configure.ac:39: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:66: error: possibly undefined macro: AC_FL_STATIC
configure.ac:71: error: possibly undefined macro: AC_FL_OPT

**Reply by dl1ycf** [reply]
can you tell me the version of autoconf? Here I have version 2.71
which should be quite recent.

I had a lot of problems with the latest version of autoconf and, for
example, had to replace all AC_HELP_STRING  to AS_HELP_STRING
and a number of other changes. BTW these are already made in the
fldigi repo, it seems that the original flwkey repo is somewhat out-dated.

The use of AM_INIT_AUTOMAKE is still the same in fldigi, so can you 
compile fldigi? Further, AC_FL_STATIC is actually defined in m4/static.m4
so did you do "autoreconf -i" to bootstrap? The usual sequence of commands
is

autoreconf -i
./configure
make

**Reply by dl1ycf** [reply]
BTW the "configure" file is left out intentionally since it might depend on the autoconf version in use. I just made a fresh "clone" on my Raspi running "buster", and "autoreconf -i" + "./configure" + "make" succeeded. On the RaspPi, autoconf is Version 2.69.
So I am a little bit lost what went wrong on your ubuntu.

**Reply by jimahlstrom** [reply]
My autoconf is 2.69. That is the most recent on Ubuntu LTS. The "autoreconf -i" worked to create the "configure" file. This disagrees with the instructions in INSTALL. It seems that distributing "configure" is a better option. But the link fails because Ubuntu creates a PIE executable by default, and libfltk is not PIE. Doctor Google says that the link option -no-pie fixes this, so I will try that next.

**Reply by jimahlstrom** [reply]
Problem solved. I needed to remove an old install of libfltk in /usr/local/lib. Could you create a "configure" in your git directory?

**Comment by dl1ycf** [comment]
I added a "configure" file. Just checked that the "configure" created with autoconf 2.71
also works on my RaspPi where there is autoconf 2.69.

Note that there is no "configure" file in the original master repo on SourceForge,
so "autoreconf" seems to be the method of choice, but you are right it is not in the INSTALL
instructions.


> Am 25.02.2022 um 16:17 schrieb jimahlstrom ***@***.***>:
> 
> 
> Problem solved. I needed to remove an old install of libfltk in /usr/local/lib. Could you create a "configure" in your git directory?
>

**Reply by jimahlstrom** [reply]
I can verify that your configure file works correctly on my Ubuntu 20.04.

**Comment by dl1ycf** [comment]
I did an update of the Winkey emulator, adding "buffered commands" since especially the "Set Buffered Speed" and "Cancel Buffered Speed" are often used in contests. Apart from that, I think the WinKey emulator is close to convergence now, since there must be a cut somewhere between a "light-weight" (TeensyWinKeyEmulator) and "full-fledged" (K3NG emulator) implementation.
